src/crimson/os/seastore/seastore_types.h

```diff
@@ -377,6 +378,104 @@ class segment_map_t {
+  static constexpr internal_block_id_t DEFAULT_INTERNAL_BLK_ID =
+    (std::numeric_limits<internal_block_id_t>::max() >>
+    (std::numeric_limits<internal_block_id_t>::digits - BLK_ID_LEN_BITS))
+    - 1;
+
+  internal_block_id_t block = DEFAULT_INTERNAL_BLK_ID;
+
+  constexpr block_id_t(uint64_t encoded) : block(encoded) {}
+public:
+  block_id_t() = default;
+  constexpr block_id_t(device_id_t id, device_block_id_t block)
+    : block(make_internal(block, id)) {}
+
+  [[gnu::always_inline]]
+  device_id_t device_id() const {
+    return internal_to_device(block);
+  }
+
+  [[gnu::always_inline]]
+  constexpr device_block_id_t device_block_id() const {
+    return internal_to_block(block);
+  }
+
+  bool operator==(const block_id_t& other) const {
+    return block == other.block;
+  }
+  bool operator!=(const block_id_t& other) const {
+    return block != other.block;
+  }
+  bool operator<(const block_id_t& other) const {
+    return block < other.block;
+  }
+  bool operator<=(const block_id_t& other) const {
+    return block <= other.block;
+  }
+  bool operator>(const block_id_t& other) const {
+    return block > other.block;
+  }
+  bool operator>=(const block_id_t& other) const {
+    return block >= other.block;
+  }
+
+private:
+  static constexpr unsigned block_id_bits = (
+    BLK_ID_LEN_BITS
+  );
+
+  static inline device_id_t internal_to_device(internal_block_id_t id) {
+    return (static_cast<device_id_t>(id) & BLK_ID_MASK) >> block_id_bits;
+  }
+
+  constexpr static inline device_block_id_t internal_to_block(
+    internal_block_id_t id) {
+    return id & (~BLK_ID_MASK);
+  }
+
+  constexpr static inline internal_block_id_t make_internal(
+    device_block_id_t id,
+    device_id_t d_id) {
+    return static_cast<internal_block_id_t>(id) | // block id
+      (static_cast<internal_block_id_t>(d_id) << block_id_bits) | // device id
+      (static_cast<internal_block_id_t>(0x1) << // addr_type
+      (BLK_ID_LEN_BITS + DEVICE_ID_LEN_BITS - 1));
+  }
+
+  friend struct blk_paddr_t;
+  friend struct paddr_t;
+  friend struct paddr_le_t;
+};
+
+constexpr block_id_t NULL_BLK_ID = block_id_t(DEVICE_ID_NULL, 0);
+constexpr block_id_t FAKE_BLK_ID = block_id_t(DEVICE_ID_FAKE, 0);
+
+static constexpr uint16_t BLK_OFF_LEN_BITS = 20;
+constexpr seastore_off_t BLK_OFF_MAX =
+  std::numeric_limits<seastore_off_t>::max() >>
+  (std::numeric_limits<seastore_off_t>::digits - BLK_OFF_LEN_BITS);
+constexpr seastore_off_t BLK_OFF_MIN =
+  std::numeric_limits<seastore_off_t>::min() /
+  (1 << std::numeric_limits<seastore_off_t>::digits - BLK_OFF_LEN_BITS);
+constexpr uint32_t BLK_OFF_NUM_MASK = 0x7FFFF;
+constexpr uint32_t BLK_OFF_MSB_MASK = 0x80000;
+constexpr seastore_off_t NULL_BLK_OFF = BLK_OFF_MAX;
+constexpr seastore_off_t MAX_BLK_OFF = BLK_OFF_MAX;
+static void check_blk_off_valid(const seastore_off_t offset) {
+  assert(offset < BLK_OFF_MAX);
+  assert(offset > BLK_OFF_MIN);
+}
```

Comment:
Why do we need both MAX_BLK_OFF and BLK_OFF_MAX?  Rename NULL_BLK_OFF to BLK_OFF_NULL.

Reply:
ok